crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+        );
+        let offset = total_dunder_all_range.start();
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
+                .collect(),
+            comment_range: comment_range.map(|c| c + offset),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DunderAllLine {
+    JustAComment(LineWithJustAComment),
+    OneOrMoreItems(LineWithItems),
+}
+
+fn collect_dunder_all_items(lines: &[DunderAllLine]) -> Vec<DunderAllItem> {
+    // Given data on each line in `__all__`, group lines together into "items".
+    // Each item contains exactly one element,
+    // but might contain multiple comments attached to that element
+    // that must move with the element when `__all__` is sorted.
+    let mut all_items = vec![];
+    let mut this_range = None;
+    for line in lines {
+        match line {
+            DunderAllLine::JustAComment(LineWithJustAComment(comment_range)) => {
+                this_range = Some(*comment_range);
+            }
+            DunderAllLine::OneOrMoreItems(LineWithItems {
+                items,
+                comment_range,
+            }) => {
+                let [(first_val, first_range), rest @ ..] = items.as_slice() else {
+                    unreachable!(
+                        "LineWithItems::new() should uphold the invariant that this list is always non-empty"
+                    )
+                };
+                let range = this_range.map_or(*first_range, |r| {
+                    TextRange::new(r.start(), first_range.end())
+                });
+                all_items.push(DunderAllItem {
+                    value: first_val.clone(),
+                    original_index: all_items.len(),
+                    range,
+                    additional_comments: *comment_range,
+                });
+                this_range = None;
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
+                        original_index: all_items.len(),
+                        range: *range,
+                        additional_comments: None,
+                    });
+                }
+            }
+        }
+    }
+    all_items
+}
+
+#[derive(Clone, Debug)]
+struct DunderAllItem {
+    value: String,
+    // Each `AllItem` in any given list should have a unique `original_index`:
+    original_index: usize,
+    // Note that this range might include comments, etc.
+    range: TextRange,
+    additional_comments: Option<TextRange>,
+}
+
+impl Ranged for DunderAllItem {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
+
+impl PartialEq for DunderAllItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.original_index == other.original_index
+    }
+}
+
+impl Eq for DunderAllItem {}
+
+impl DunderAllItem {
+    fn sort_index(&self) -> (&str, usize) {
+        (&self.value, self.original_index)
+    }
+}
```

Comment:
Can you explain (or add a comment) why `sort_index` is needed. My guess is that it is to preserve the original order, but that's something that Rust's `sort` method should already give us for free.

Reply:
yup, my worry here was about the possibility of an `__all__` that had duplicate items in it. I wanted to ensure that in the eventuality where two items in the list compared equal, the item that was originally first in the list remained first in the list. That invariant will hold even if I just use `self.value` as the key for `.sort()`?